src/protocols/secure_channel/CASESession.cpp

```diff
@@ -683,64 +667,42 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
 }
 
 #if INET_CONFIG_ENABLE_TCP_ENDPOINT
-void CASESession::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR err)
+void CASESession::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionHolder & conn, CHIP_ERROR err)
 {
-    VerifyOrReturn(conn != nullptr);
-    // conn->mAppState should not be NULL. SessionManager has already checked
-    // before calling this callback.
-    VerifyOrDie(conn->mAppState != nullptr);
+    VerifyOrReturn(conn == mPeerConnState);
 
     char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
     conn->mPeerAddr.ToString(peerAddrBuf);
 
-    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
-    VerifyOrReturn(caseSession != nullptr);
+    auto ConnectionCleanup = [](CASESession * s) {
+        s->mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->ReleaseTCPConnection();
+        s->mSecureSessionHolder.Get().Value()->AsSecureSession()->ReleaseTCPConnection();
+        s->Clear();
+    };
+    std::unique_ptr<CASESession, decltype(ConnectionCleanup)> connectionHolder(this, ConnectionCleanup);
 
-    // Exit and disconnect if connection setup encountered an error.
-    SuccessOrExit(err);
+    // Bail if connection setup encountered an error.
+    LogAndReturnOnFailure(err, SecureChannel, "Connection establishment failed with peer at %s", peerAddrBuf);
 
     ChipLogDetail(SecureChannel, "TCP Connection established with %s before session establishment", peerAddrBuf);
 
     // Associate the connection with the current unauthenticated session for the
     // CASE exchange.
-    caseSession->mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
+    mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->SetTCPConnection(conn);
 
     // Associate the connection with the current secure session that is being
     // set up.
-    caseSession->mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
+    mSecureSessionHolder.Get().Value()->AsSecureSession()->SetTCPConnection(conn);
 
     // Send Sigma1 after connection is established for sessions over TCP
-    err = caseSession->SendSigma1();
-    SuccessOrExit(err);
+    LogAndReturnOnFailure(SendSigma1(), SecureChannel, "Sigma1 failed to peer %s", peerAddrBuf);
```

Comment:
I have called this out in the follow up PR so I'll just leave this here for reference: I think we might need to propagate the error in case of these failures which was missed previously as well but the refactor of cleanup into a closure makes it tricky to time the cleanup before error propagation: https://github.com/project-chip/connectedhomeip/pull/40904#discussion_r2337896088